clientv3/integration: TestTxnWriteFail

```
=== RUN   TestTxnWriteFail
2016-05-20 06:05:40.712844 I | etcdserver: name = node2727028312981969055
2016-05-20 06:05:40.712954 I | etcdserver: data dir = /tmp/etcd551644292
2016-05-20 06:05:40.713065 I | etcdserver: member dir = /tmp/etcd551644292/member
2016-05-20 06:05:40.713129 I | etcdserver: heartbeat = 10ms
2016-05-20 06:05:40.713199 I | etcdserver: election = 100ms
2016-05-20 06:05:40.713284 I | etcdserver: snapshot count = 0
2016-05-20 06:05:40.713362 I | etcdserver: advertise client URLs = http://127.0.0.1:21532
2016-05-20 06:05:40.713470 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21531
2016-05-20 06:05:40.713543 I | etcdserver: name = node3807024210026896221
2016-05-20 06:05:40.713675 I | etcdserver: data dir = /tmp/etcd375866387
2016-05-20 06:05:40.713758 I | etcdserver: member dir = /tmp/etcd375866387/member
2016-05-20 06:05:40.713861 I | etcdserver: name = node8889113302011076440
2016-05-20 06:05:40.713989 I | etcdserver: data dir = /tmp/etcd550152278
2016-05-20 06:05:40.714083 I | etcdserver: member dir = /tmp/etcd550152278/member
2016-05-20 06:05:40.714165 I | etcdserver: heartbeat = 10ms
2016-05-20 06:05:40.714258 I | etcdserver: heartbeat = 10ms
2016-05-20 06:05:40.714356 I | etcdserver: election = 100ms
2016-05-20 06:05:40.714454 I | etcdserver: snapshot count = 0
2016-05-20 06:05:40.714549 I | etcdserver: initial cluster = node2727028312981969055=http://127.0.0.1:21531,node3807024210026896221=http://127.0.0.1:21533,node8889113302011076440=http://127.0.0.1:21535
2016-05-20 06:05:40.714694 I | etcdserver: election = 100ms
2016-05-20 06:05:40.714800 I | etcdserver: snapshot count = 0
2016-05-20 06:05:40.714926 I | etcdserver: advertise client URLs = http://127.0.0.1:21536
2016-05-20 06:05:40.715033 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21535
2016-05-20 06:05:40.715186 I | etcdserver: advertise client URLs = http://127.0.0.1:21534
2016-05-20 06:05:40.715342 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21533
2016-05-20 06:05:40.715490 I | etcdserver: initial cluster = node2727028312981969055=http://127.0.0.1:21531,node3807024210026896221=http://127.0.0.1:21533,node8889113302011076440=http://127.0.0.1:21535
2016-05-20 06:05:40.715625 I | etcdserver: initial cluster = node2727028312981969055=http://127.0.0.1:21531,node3807024210026896221=http://127.0.0.1:21533,node8889113302011076440=http://127.0.0.1:21535
2016-05-20 06:05:40.752477 I | etcdserver: starting member 6fac80429b1976f1 in cluster d909c0d7b220e389
2016-05-20 06:05:40.752782 I | raft: 6fac80429b1976f1 became follower at term 0
2016-05-20 06:05:40.752865 I | raft: newRaft 6fac80429b1976f1 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-20 06:05:40.752915 I | raft: 6fac80429b1976f1 became follower at term 1
2016-05-20 06:05:40.752993 I | etcdserver: starting member e43be156984881f7 in cluster d909c0d7b220e389
2016-05-20 06:05:40.753215 I | raft: e43be156984881f7 became follower at term 0
2016-05-20 06:05:40.753306 I | raft: newRaft e43be156984881f7 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-20 06:05:40.753363 I | raft: e43be156984881f7 became follower at term 1
2016-05-20 06:05:40.761315 I | etcdserver: starting member aea1f3ff542bdddf in cluster d909c0d7b220e389
2016-05-20 06:05:40.761671 I | raft: aea1f3ff542bdddf became follower at term 0
2016-05-20 06:05:40.761862 I | raft: newRaft aea1f3ff542bdddf [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-20 06:05:40.761993 I | raft: aea1f3ff542bdddf became follower at term 1
2016-05-20 06:05:40.784310 I | etcdserver: set snapshot count to default 10000
2016-05-20 06:05:40.784490 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-20 06:05:40.790548 I | etcdserver: set snapshot count to default 10000
2016-05-20 06:05:40.790820 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-20 06:05:40.793215 N | etcdserver: added local member 6fac80429b1976f1 [http://127.0.0.1:21531] to cluster d909c0d7b220e389
2016-05-20 06:05:40.793997 N | etcdserver: added member aea1f3ff542bdddf [http://127.0.0.1:21535] to cluster d909c0d7b220e389
2016-05-20 06:05:40.795219 N | etcdserver: added member e43be156984881f7 [http://127.0.0.1:21533] to cluster d909c0d7b220e389
2016-05-20 06:05:40.800231 I | rafthttp: the connection with e43be156984881f7 became active
2016-05-20 06:05:40.801097 I | rafthttp: the connection with 6fac80429b1976f1 became active
2016-05-20 06:05:40.805315 N | etcdserver: added member 6fac80429b1976f1 [http://127.0.0.1:21531] to cluster d909c0d7b220e389
2016-05-20 06:05:40.805648 I | rafthttp: the connection with aea1f3ff542bdddf became active
2016-05-20 06:05:40.805748 I | etcdserver: set snapshot count to default 10000
2016-05-20 06:05:40.806592 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-20 06:05:40.806818 I | rafthttp: the connection with 6fac80429b1976f1 became active
2016-05-20 06:05:40.808301 N | etcdserver: added member aea1f3ff542bdddf [http://127.0.0.1:21535] to cluster d909c0d7b220e389
2016-05-20 06:05:40.809172 N | etcdserver: added local member e43be156984881f7 [http://127.0.0.1:21533] to cluster d909c0d7b220e389
2016-05-20 06:05:40.809874 N | etcdserver: added member 6fac80429b1976f1 [http://127.0.0.1:21531] to cluster d909c0d7b220e389
2016-05-20 06:05:40.810303 N | etcdserver: added local member aea1f3ff542bdddf [http://127.0.0.1:21535] to cluster d909c0d7b220e389
2016-05-20 06:05:40.810741 N | etcdserver: added member e43be156984881f7 [http://127.0.0.1:21533] to cluster d909c0d7b220e389
2016-05-20 06:05:40.811601 I | rafthttp: the connection with aea1f3ff542bdddf became active
2016-05-20 06:05:40.813333 I | rafthttp: the connection with e43be156984881f7 became active
2016-05-20 06:05:40.843427 I | raft: 6fac80429b1976f1 is starting a new election at term 1
2016-05-20 06:05:40.843549 I | raft: 6fac80429b1976f1 became candidate at term 2
2016-05-20 06:05:40.843612 I | raft: 6fac80429b1976f1 received vote from 6fac80429b1976f1 at term 2
2016-05-20 06:05:40.843685 I | raft: 6fac80429b1976f1 [logterm: 1, index: 3] sent vote request to aea1f3ff542bdddf at term 2
2016-05-20 06:05:40.843754 I | raft: 6fac80429b1976f1 [logterm: 1, index: 3] sent vote request to e43be156984881f7 at term 2
2016-05-20 06:05:40.844951 I | raft: aea1f3ff542bdddf [term: 1] received a MsgVote message with higher term from 6fac80429b1976f1 [term: 2]
2016-05-20 06:05:40.845043 I | raft: e43be156984881f7 [term: 1] received a MsgVote message with higher term from 6fac80429b1976f1 [term: 2]
2016-05-20 06:05:40.845192 I | raft: e43be156984881f7 became follower at term 2
2016-05-20 06:05:40.845268 I | raft: aea1f3ff542bdddf became follower at term 2
2016-05-20 06:05:40.845330 I | raft: e43be156984881f7 [logterm: 1, index: 3, vote: 0] voted for 6fac80429b1976f1 [logterm: 1, index: 3] at term 2
2016-05-20 06:05:40.845416 I | raft: aea1f3ff542bdddf [logterm: 1, index: 3, vote: 0] voted for 6fac80429b1976f1 [logterm: 1, index: 3] at term 2
2016-05-20 06:05:40.845957 I | raft: 6fac80429b1976f1 received vote from aea1f3ff542bdddf at term 2
2016-05-20 06:05:40.846021 I | raft: 6fac80429b1976f1 [quorum:2] has received 2 votes and 0 vote rejections
2016-05-20 06:05:40.846095 I | raft: 6fac80429b1976f1 became leader at term 2
2016-05-20 06:05:40.846158 I | raft: raft.node: 6fac80429b1976f1 elected leader 6fac80429b1976f1 at term 2
2016-05-20 06:05:40.846695 I | raft: raft.node: aea1f3ff542bdddf elected leader 6fac80429b1976f1 at term 2
2016-05-20 06:05:40.847134 I | raft: raft.node: e43be156984881f7 elected leader 6fac80429b1976f1 at term 2
2016-05-20 06:05:40.849514 I | etcdserver: published {Name:node2727028312981969055 ClientURLs:[http://127.0.0.1:21532]} to cluster d909c0d7b220e389
2016-05-20 06:05:40.850589 I | etcdserver: published {Name:node8889113302011076440 ClientURLs:[http://127.0.0.1:21536]} to cluster d909c0d7b220e389
2016-05-20 06:05:40.850991 I | etcdserver: published {Name:node3807024210026896221 ClientURLs:[http://127.0.0.1:21534]} to cluster d909c0d7b220e389
2016-05-20 06:05:40.851059 I | etcdserver: setting up the initial cluster version to 3.0
2016-05-20 06:05:40.851975 N | membership: set the initial cluster version to 3.0
2016-05-20 06:05:40.852442 N | membership: set the initial cluster version to 3.0
2016-05-20 06:05:40.852597 N | membership: set the initial cluster version to 3.0
2016-05-20 06:05:40.868305 E | rafthttp: failed to read aea1f3ff542bdddf on stream Message (net/http: request canceled)
2016-05-20 06:05:40.868385 I | rafthttp: the connection with aea1f3ff542bdddf became inactive
2016-05-20 06:05:40.953143 I | raft: aea1f3ff542bdddf is starting a new election at term 2
2016-05-20 06:05:40.953352 I | raft: aea1f3ff542bdddf became candidate at term 3
2016-05-20 06:05:40.953458 I | raft: aea1f3ff542bdddf received vote from aea1f3ff542bdddf at term 3
2016-05-20 06:05:40.953606 I | raft: aea1f3ff542bdddf [logterm: 2, index: 8] sent vote request to 6fac80429b1976f1 at term 3
2016-05-20 06:05:40.953747 I | raft: aea1f3ff542bdddf [logterm: 2, index: 8] sent vote request to e43be156984881f7 at term 3
2016-05-20 06:05:40.953851 I | raft: raft.node: aea1f3ff542bdddf lost leader 6fac80429b1976f1 at term 3
2016-05-20 06:05:40.954786 I | raft: e43be156984881f7 [term: 2] received a MsgVote message with higher term from aea1f3ff542bdddf [term: 3]
2016-05-20 06:05:40.954943 I | raft: e43be156984881f7 became follower at term 3
2016-05-20 06:05:40.955097 I | raft: e43be156984881f7 [logterm: 2, index: 8, vote: 0] voted for aea1f3ff542bdddf [logterm: 2, index: 8] at term 3
2016-05-20 06:05:40.955206 I | raft: raft.node: e43be156984881f7 lost leader 6fac80429b1976f1 at term 3
2016-05-20 06:05:40.955894 I | raft: aea1f3ff542bdddf received vote from e43be156984881f7 at term 3
2016-05-20 06:05:40.955994 I | raft: aea1f3ff542bdddf [quorum:2] has received 2 votes and 0 vote rejections
2016-05-20 06:05:40.956122 I | raft: aea1f3ff542bdddf became leader at term 3
2016-05-20 06:05:40.956243 I | raft: raft.node: aea1f3ff542bdddf elected leader aea1f3ff542bdddf at term 3
2016-05-20 06:05:40.957164 I | raft: raft.node: e43be156984881f7 elected leader aea1f3ff542bdddf at term 3
2016-05-20 06:05:40.958724 E | rafthttp: failed to dial 6fac80429b1976f1 on stream MsgApp v2 (dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:40.958848 I | rafthttp: the connection with 6fac80429b1976f1 became inactive
2016-05-20 06:05:40.958978 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21531) of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:40.959147 W | etcdserver: cannot get the version of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:40.972053 E | rafthttp: failed to dial 6fac80429b1976f1 on stream MsgApp v2 (dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:40.972214 I | rafthttp: the connection with 6fac80429b1976f1 became inactive
2016-05-20 06:05:41.293534 I | v2http: enabled capabilities for version 3.0.0
2016-05-20 06:05:41.311396 I | v2http: enabled capabilities for version 3.0.0
2016-05-20 06:05:44.963000 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21531) of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:44.963277 W | etcdserver: cannot get the version of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:48.967789 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21531) of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:48.967959 W | etcdserver: cannot get the version of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:52.970492 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21531) of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:52.970736 W | etcdserver: cannot get the version of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
--- FAIL: TestTxnWriteFail (12.42s)
	txn_test.go:78: timed out waiting for txn to fail
```

@heyitsanthony Thanks!

Pasting here:

```
=== RUN   TestTxnWriteFail
2016-05-28 00:32:48.493269 I | integration: launching node4890084730240107037 (unix://localhost:node4890084730240107037.sock)
2016-05-28 00:32:48.494095 I | integration: launching node5637448722635423475 (unix://localhost:node5637448722635423475.sock)
2016-05-28 00:32:48.494473 I | integration: launching node2216460087929149329 (unix://localhost:node2216460087929149329.sock)
2016-05-28 00:32:48.514269 I | etcdserver: name = node4890084730240107037
2016-05-28 00:32:48.514320 I | etcdserver: data dir = /tmp/etcd513148213
2016-05-28 00:32:48.514356 I | etcdserver: member dir = /tmp/etcd513148213/member
2016-05-28 00:32:48.514385 I | etcdserver: heartbeat = 10ms
2016-05-28 00:32:48.514409 I | etcdserver: election = 100ms
2016-05-28 00:32:48.514435 I | etcdserver: snapshot count = 0
2016-05-28 00:32:48.514467 I | etcdserver: advertise client URLs = http://127.0.0.1:21326
2016-05-28 00:32:48.514495 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21325
2016-05-28 00:32:48.514551 I | etcdserver: initial cluster = node2216460087929149329=http://127.0.0.1:21327,node4890084730240107037=http://127.0.0.1:21325,node5637448722635423475=http://127.0.0.1:21323
2016-05-28 00:32:48.514865 I | etcdserver: name = node5637448722635423475
2016-05-28 00:32:48.514907 I | etcdserver: data dir = /tmp/etcd956283822
2016-05-28 00:32:48.514946 I | etcdserver: member dir = /tmp/etcd956283822/member
2016-05-28 00:32:48.514974 I | etcdserver: heartbeat = 10ms
2016-05-28 00:32:48.515002 I | etcdserver: election = 100ms
2016-05-28 00:32:48.515029 I | etcdserver: snapshot count = 0
2016-05-28 00:32:48.515063 I | etcdserver: advertise client URLs = http://127.0.0.1:21324
2016-05-28 00:32:48.515098 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21323
2016-05-28 00:32:48.515159 I | etcdserver: initial cluster = node2216460087929149329=http://127.0.0.1:21327,node4890084730240107037=http://127.0.0.1:21325,node5637448722635423475=http://127.0.0.1:21323
2016-05-28 00:32:48.515256 I | etcdserver: name = node2216460087929149329
2016-05-28 00:32:48.515297 I | etcdserver: data dir = /tmp/etcd141558544
2016-05-28 00:32:48.515332 I | etcdserver: member dir = /tmp/etcd141558544/member
2016-05-28 00:32:48.515359 I | etcdserver: heartbeat = 10ms
2016-05-28 00:32:48.515386 I | etcdserver: election = 100ms
2016-05-28 00:32:48.515408 I | etcdserver: snapshot count = 0
2016-05-28 00:32:48.515455 I | etcdserver: advertise client URLs = http://127.0.0.1:21328
2016-05-28 00:32:48.515485 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21327
2016-05-28 00:32:48.515540 I | etcdserver: initial cluster = node2216460087929149329=http://127.0.0.1:21327,node4890084730240107037=http://127.0.0.1:21325,node5637448722635423475=http://127.0.0.1:21323
2016-05-28 00:32:48.538407 I | etcdserver: starting member 2f60caad9ce091e9 in cluster 15f3f7546f78db91
2016-05-28 00:32:48.538602 I | raft: 2f60caad9ce091e9 became follower at term 0
2016-05-28 00:32:48.538673 I | raft: newRaft 2f60caad9ce091e9 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-28 00:32:48.538722 I | raft: 2f60caad9ce091e9 became follower at term 1
2016-05-28 00:32:48.539685 I | etcdserver: starting member e669529ac5ff32e7 in cluster 15f3f7546f78db91
2016-05-28 00:32:48.539837 I | raft: e669529ac5ff32e7 became follower at term 0
2016-05-28 00:32:48.539907 I | raft: newRaft e669529ac5ff32e7 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-28 00:32:48.539947 I | raft: e669529ac5ff32e7 became follower at term 1
2016-05-28 00:32:48.541055 I | etcdserver: starting member 5281ddf5029be584 in cluster 15f3f7546f78db91
2016-05-28 00:32:48.541188 I | raft: 5281ddf5029be584 became follower at term 0
2016-05-28 00:32:48.541234 I | raft: newRaft 5281ddf5029be584 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-28 00:32:48.541262 I | raft: 5281ddf5029be584 became follower at term 1
2016-05-28 00:32:48.568630 I | etcdserver: set snapshot count to default 10000
2016-05-28 00:32:48.568701 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-28 00:32:48.573091 N | etcdserver: added local member 2f60caad9ce091e9 [http://127.0.0.1:21325] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.573470 N | etcdserver: added member 5281ddf5029be584 [http://127.0.0.1:21323] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.573810 N | etcdserver: added member e669529ac5ff32e7 [http://127.0.0.1:21327] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.577390 I | etcdserver: set snapshot count to default 10000
2016-05-28 00:32:48.577452 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-28 00:32:48.584061 I | etcdserver: set snapshot count to default 10000
2016-05-28 00:32:48.584142 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-28 00:32:48.584931 I | rafthttp: the connection with 2f60caad9ce091e9 became active
2016-05-28 00:32:48.586208 N | etcdserver: added member 2f60caad9ce091e9 [http://127.0.0.1:21325] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.587513 I | integration: launched node5637448722635423475 (unix://localhost:node5637448722635423475.sock)
2016-05-28 00:32:48.587843 N | etcdserver: added local member 5281ddf5029be584 [http://127.0.0.1:21323] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.588455 N | etcdserver: added member e669529ac5ff32e7 [http://127.0.0.1:21327] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.589929 I | rafthttp: the connection with 2f60caad9ce091e9 became active
2016-05-28 00:32:48.591135 N | etcdserver: added member 2f60caad9ce091e9 [http://127.0.0.1:21325] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.591342 I | integration: launched node4890084730240107037 (unix://localhost:node4890084730240107037.sock)
2016-05-28 00:32:48.591722 N | etcdserver: added member 5281ddf5029be584 [http://127.0.0.1:21323] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.592325 N | etcdserver: added local member e669529ac5ff32e7 [http://127.0.0.1:21327] to cluster 15f3f7546f78db91
2016-05-28 00:32:48.592982 I | integration: launched node2216460087929149329 (unix://localhost:node2216460087929149329.sock)
2016-05-28 00:32:48.594871 I | rafthttp: the connection with 5281ddf5029be584 became active
2016-05-28 00:32:48.599024 I | rafthttp: the connection with e669529ac5ff32e7 became active
2016-05-28 00:32:48.600279 I | rafthttp: the connection with e669529ac5ff32e7 became active
2016-05-28 00:32:48.601239 I | rafthttp: the connection with 5281ddf5029be584 became active
2016-05-28 00:32:48.630687 I | raft: 2f60caad9ce091e9 is starting a new election at term 1
2016-05-28 00:32:48.630778 I | raft: 2f60caad9ce091e9 became candidate at term 2
2016-05-28 00:32:48.630831 I | raft: 2f60caad9ce091e9 received vote from 2f60caad9ce091e9 at term 2
2016-05-28 00:32:48.630894 I | raft: 2f60caad9ce091e9 [logterm: 1, index: 3] sent vote request to 5281ddf5029be584 at term 2
2016-05-28 00:32:48.630948 I | raft: 2f60caad9ce091e9 [logterm: 1, index: 3] sent vote request to e669529ac5ff32e7 at term 2
2016-05-28 00:32:48.631491 I | raft: 5281ddf5029be584 [term: 1] received a MsgVote message with higher term from 2f60caad9ce091e9 [term: 2]
2016-05-28 00:32:48.631607 I | raft: 5281ddf5029be584 became follower at term 2
2016-05-28 00:32:48.631673 I | raft: 5281ddf5029be584 [logterm: 1, index: 3, vote: 0] voted for 2f60caad9ce091e9 [logterm: 1, index: 3] at term 2
2016-05-28 00:32:48.631721 I | raft: e669529ac5ff32e7 [term: 1] received a MsgVote message with higher term from 2f60caad9ce091e9 [term: 2]
2016-05-28 00:32:48.631876 I | raft: e669529ac5ff32e7 became follower at term 2
2016-05-28 00:32:48.631978 I | raft: e669529ac5ff32e7 [logterm: 1, index: 3, vote: 0] voted for 2f60caad9ce091e9 [logterm: 1, index: 3] at term 2
2016-05-28 00:32:48.632334 I | raft: 2f60caad9ce091e9 received vote from 5281ddf5029be584 at term 2
2016-05-28 00:32:48.632427 I | raft: 2f60caad9ce091e9 [quorum:2] has received 2 votes and 0 vote rejections
2016-05-28 00:32:48.632515 I | raft: 2f60caad9ce091e9 became leader at term 2
2016-05-28 00:32:48.632588 I | raft: raft.node: 2f60caad9ce091e9 elected leader 2f60caad9ce091e9 at term 2
2016-05-28 00:32:48.633144 I | raft: raft.node: 5281ddf5029be584 elected leader 2f60caad9ce091e9 at term 2
2016-05-28 00:32:48.633228 I | raft: raft.node: e669529ac5ff32e7 elected leader 2f60caad9ce091e9 at term 2
2016-05-28 00:32:48.637481 I | etcdserver: published {Name:node4890084730240107037 ClientURLs:[http://127.0.0.1:21326]} to cluster 15f3f7546f78db91
2016-05-28 00:32:48.639986 I | etcdserver: published {Name:node2216460087929149329 ClientURLs:[http://127.0.0.1:21328]} to cluster 15f3f7546f78db91
2016-05-28 00:32:48.641791 I | etcdserver: published {Name:node5637448722635423475 ClientURLs:[http://127.0.0.1:21324]} to cluster 15f3f7546f78db91
2016-05-28 00:32:48.643606 I | etcdserver: setting up the initial cluster version to 3.0
2016-05-28 00:32:48.644916 N | membership: set the initial cluster version to 3.0
2016-05-28 00:32:48.645609 N | membership: set the initial cluster version to 3.0
2016-05-28 00:32:48.645681 N | membership: set the initial cluster version to 3.0
2016-05-28 00:32:48.659581 I | integration: stopping node5637448722635423475 (unix://localhost:node5637448722635423475.sock)
2016-05-28 00:32:48.662768 E | rafthttp: failed to read 2f60caad9ce091e9 on stream MsgApp v2 (net/http: request canceled)
2016-05-28 00:32:48.662867 I | rafthttp: the connection with 2f60caad9ce091e9 became inactive
2016-05-28 00:32:48.686720 W | etcdserver: failed to send out heartbeat on time (exceeded the 10ms timeout for 5.775802ms)
2016-05-28 00:32:48.686785 W | etcdserver: server is likely overloaded
2016-05-28 00:32:48.686826 W | etcdserver: failed to send out heartbeat on time (exceeded the 10ms timeout for 5.905021ms)
2016-05-28 00:32:48.686851 W | etcdserver: server is likely overloaded
2016-05-28 00:32:48.687736 E | rafthttp: failed to read e669529ac5ff32e7 on stream Message (net/http: request canceled)
2016-05-28 00:32:48.687791 I | rafthttp: the connection with e669529ac5ff32e7 became inactive
2016-05-28 00:32:48.693654 I | integration: stopped node5637448722635423475 (unix://localhost:node5637448722635423475.sock)
2016-05-28 00:32:48.710837 E | rafthttp: failed to write 5281ddf5029be584 on stream Message (write tcp 127.0.0.1:21325->127.0.0.1:59411: write: broken pipe)
2016-05-28 00:32:48.710893 I | rafthttp: the connection with 5281ddf5029be584 became inactive
2016-05-28 00:32:48.764746 E | rafthttp: failed to dial 5281ddf5029be584 on stream MsgApp v2 (dial tcp 127.0.0.1:21323: getsockopt: connection refused)
2016-05-28 00:32:48.764843 I | rafthttp: the connection with 5281ddf5029be584 became inactive
2016-05-28 00:32:52.645064 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21323) of member 5281ddf5029be584 (Get http://127.0.0.1:21323/version: dial tcp 127.0.0.1:21323: getsockopt: connection refused)
2016-05-28 00:32:52.645337 W | etcdserver: cannot get the version of member 5281ddf5029be584 (Get http://127.0.0.1:21323/version: dial tcp 127.0.0.1:21323: getsockopt: connection refused)
2016-05-28 00:32:56.648531 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21323) of member 5281ddf5029be584 (Get http://127.0.0.1:21323/version: dial tcp 127.0.0.1:21323: getsockopt: connection refused)
2016-05-28 00:32:56.648755 W | etcdserver: cannot get the version of member 5281ddf5029be584 (Get http://127.0.0.1:21323/version: dial tcp 127.0.0.1:21323: getsockopt: connection refused)
2016-05-28 00:33:00.653297 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21323) of member 5281ddf5029be584 (Get http://127.0.0.1:21323/version: dial tcp 127.0.0.1:21323: getsockopt: connection refused)
2016-05-28 00:33:00.653580 W | etcdserver: cannot get the version of member 5281ddf5029be584 (Get http://127.0.0.1:21323/version: dial tcp 127.0.0.1:21323: getsockopt: connection refused)
2016-05-28 00:33:00.816558 I | integration: terminating node5637448722635423475 (unix://localhost:node5637448722635423475.sock)
2016-05-28 00:33:00.835251 I | integration: terminated node5637448722635423475 (unix://localhost:node5637448722635423475.sock)
2016-05-28 00:33:00.835395 I | integration: terminating node4890084730240107037 (unix://localhost:node4890084730240107037.sock)
2016-05-28 00:33:00.840682 E | rafthttp: failed to read e669529ac5ff32e7 on stream Message (net/http: request canceled)
2016-05-28 00:33:00.840831 I | rafthttp: the connection with e669529ac5ff32e7 became inactive
2016-05-28 00:33:00.875694 I | integration: terminated node4890084730240107037 (unix://localhost:node4890084730240107037.sock)
2016-05-28 00:33:00.875841 I | integration: terminating node2216460087929149329 (unix://localhost:node2216460087929149329.sock)
2016-05-28 00:33:00.906351 I | integration: terminated node2216460087929149329 (unix://localhost:node2216460087929149329.sock)
--- FAIL: TestTxnWriteFail (12.42s)
	txn_test.go:78: timed out waiting for txn to fail
```

--- a/clientv3/client.go
+++ b/clientv3/client.go
@@ -292,6 +292,9 @@ func newClient(cfg *Config) (*Client, error) {
 func (c *Client) ActiveConnection() *grpc.ClientConn {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
+	if c.conn == nil {
+		panic("trying to return nil active connection")
+	}
 	return c.conn
 }
 

--- a/clientv3/client.go
+++ b/clientv3/client.go
@@ -101,25 +101,39 @@ func NewFromConfigFile(path string) (*Client, error) {
 }
 
 // Close shuts down the client's etcd connections.
-func (c *Client) Close() error {
+func (c *Client) Close() (err error) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
+
+	// acquire the cancel
 	if c.cancel == nil {
-		return nil
+		// already canceled
+		if c.lastConnErr != c.ctx.Err() {
+			err = c.lastConnErr
+		}
+		return
 	}
-	c.cancel()
+	cancel := c.cancel
 	c.cancel = nil
-	connc := c.newconnc
 	c.mu.Unlock()
-	c.connStartRetry(nil)
+
+	// close watcher and lease before terminating connection
+	// so they don't retry on a closed client
 	c.Watcher.Close()
 	c.Lease.Close()
+
+	// cancel reconnection loop
+	cancel()
+	c.mu.Lock()
+	connc := c.newconnc
+	c.mu.Unlock()
+	// connc on cancel() is left closed
 	<-connc
 	c.mu.Lock()
 	if c.lastConnErr != c.ctx.Err() {
-		return c.lastConnErr
+		err = c.lastConnErr
 	}
-	return nil
+	return
 }
 
 // Ctx is a context for "out of band" messages (e.g., for sending
@@ -282,30 +296,41 @@ func (c *Client) ActiveConnection() *grpc.ClientConn {
 }
 
 // retryConnection establishes a new connection
-func (c *Client) retryConnection(err error) (newConn *grpc.ClientConn, dialErr error) {
+func (c *Client) retryConnection(err error) {
+	oldconn := c.conn
+
+	// return holding lock so old connection can be cleaned up in this defer
+	defer func() {
+		if oldconn != nil {
+			oldconn.Close()
+			if st, _ := oldconn.State(); st != grpc.Shutdown {
+				// wait so grpc doesn't leak sleeping goroutines
+				oldconn.WaitForStateChange(context.Background(), st)
+			}
+		}
+		c.mu.Unlock()
+	}()
+
 	c.mu.Lock()
-	defer c.mu.Unlock()
 	if err != nil {
 		c.errors = append(c.errors, err)
 	}
-	if c.conn != nil {
-		c.conn.Close()
-		if st, _ := c.conn.State(); st != grpc.Shutdown {
-			// wait so grpc doesn't leak sleeping goroutines
-			c.conn.WaitForStateChange(context.Background(), st)
-		}
-		c.conn = nil
-	}
 	if c.cancel == nil {
 		// client has called Close() so don't try to dial out
-		return nil, c.ctx.Err()
+		return
 	}
+	c.mu.Unlock()
 
-	c.conn, dialErr = c.cfg.retryDialer(c)
+	nc, dialErr := c.cfg.retryDialer(c)
+
+	c.mu.Lock()
+	if nc != nil {
+		c.conn = nc
+	}
 	if dialErr != nil {
 		c.errors = append(c.errors, dialErr)
 	}
-	return c.conn, dialErr
+	c.lastConnErr = dialErr
 }
 
 // connStartRetry schedules a reconnect if one is not already running
@@ -321,17 +346,20 @@ func (c *Client) connStartRetry(err error) {
 
 // connWait waits for a reconnect to be processed
 func (c *Client) connWait(ctx context.Context, err error) (*grpc.ClientConn, error) {
-	c.mu.Lock()
+	c.mu.RLock()
 	ch := c.newconnc
-	c.mu.Unlock()
+	c.mu.RUnlock()
 	c.connStartRetry(err)
 	select {
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	case <-ch:
 	}
-	c.mu.Lock()
-	defer c.mu.Unlock()
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	if c.cancel == nil {
+		return c.conn, rpctypes.ErrConnClosed
+	}
 	return c.conn, c.lastConnErr
 }
 
@@ -340,11 +368,8 @@ func (c *Client) connMonitor() {
 	var err error
 
 	defer func() {
-		_, err = c.retryConnection(c.ctx.Err())
-		c.mu.Lock()
-		c.lastConnErr = err
+		c.retryConnection(c.ctx.Err())
 		close(c.newconnc)
-		c.mu.Unlock()
 	}()
 
 	limiter := rate.NewLimiter(rate.Every(minConnRetryWait), 1)
@@ -354,10 +379,8 @@ func (c *Client) connMonitor() {
 		case <-c.ctx.Done():
 			return
 		}
-		conn, connErr := c.retryConnection(err)
+		c.retryConnection(err)
 		c.mu.Lock()
-		c.lastConnErr = connErr
-		c.conn = conn
 		close(c.newconnc)
 		c.newconnc = make(chan struct{})
 		c.reconnc = make(chan error, 1)

--- a/clientv3/integration/kv_test.go
+++ b/clientv3/integration/kv_test.go
@@ -131,6 +131,13 @@ func TestKVPutWithRequireLeader(t *testing.T) {
 	if err != rpctypes.ErrNoLeader {
 		t.Fatal(err)
 	}
+
+	// clients may give timeout errors since the members are stopped; take
+	// the clients so that terminating the cluster won't complain
+	clus.Client(1).Close()
+	clus.Client(2).Close()
+	clus.TakeClient(1)
+	clus.TakeClient(2)
 }
 
 func TestKVRange(t *testing.T) {
@@ -633,13 +640,22 @@ func TestKVPutStoppedServerAndClose(t *testing.T) {
 	defer testutil.AfterTest(t)
 	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 1})
 	defer clus.Terminate(t)
+	cli := clus.Client(0)
 	clus.Members[0].Stop(t)
 	// this Put fails and triggers an asynchronous connection retry
-	_, err := clus.Client(0).Put(context.TODO(), "abc", "123")
+	_, err := cli.Put(context.TODO(), "abc", "123")
 	if err == nil ||
 		(!strings.Contains(err.Error(), "connection is closing") &&
 			!strings.Contains(err.Error(), "transport is closing")) {
 		t.Fatal(err)
 	}
-	// cluster will terminate and close the client with the retry in-flight
+
+	// wait some so the client closes with the retry in-flight
+	time.Sleep(time.Second)
+
+	// get the timeout
+	clus.TakeClient(0)
+	if err := cli.Close(); err == nil || !strings.Contains(err.Error(), "timed out") {
+		t.Fatal(err)
+	}
 }

--- a/clientv3/integration/txn_test.go
+++ b/clientv3/integration/txn_test.go
@@ -74,7 +74,7 @@ func TestTxnWriteFail(t *testing.T) {
 
 	dialTimeout := 5 * time.Second
 	select {
-	case <-time.After(2*dialTimeout + time.Second):
+	case <-time.After(dialTimeout + time.Second):
 		t.Fatalf("timed out waiting for txn to fail")
 	case <-donec:
 		// don't restart cluster until txn errors out

--- a/clientv3/remote_client.go
+++ b/clientv3/remote_client.go
@@ -88,9 +88,11 @@ func (r *remoteClient) acquire(ctx context.Context) error {
 		r.client.mu.RLock()
 		closed := r.client.cancel == nil
 		c := r.client.conn
+		lastConnErr := r.client.lastConnErr
 		match := r.conn == c
 		r.mu.Unlock()
-		if c != nil && match {
+		if lastConnErr == nil && match {
+			// new connection already
 			return nil
 		}
 		r.client.mu.RUnlock()

--- a/clientv3/txn_test.go
+++ b/clientv3/txn_test.go
@@ -20,7 +20,7 @@ import (
 )
 
 func TestTxnPanics(t *testing.T) {
-	kv := NewKV(&Client{})
+	kv := &kv{}
 
 	errc := make(chan string)
 	df := func() {

--- a/clientv3/watch.go
+++ b/clientv3/watch.go
@@ -521,6 +521,9 @@ func (w *watcher) openWatchClient() (ws pb.Watch_WatchClient, err error) {
 			return nil, v3rpc.Error(err)
 		}
 		w.rc.release()
+		if nerr := w.rc.reconnectWait(w.ctx, err); nerr != nil {
+			return nil, v3rpc.Error(nerr)
+		}
 	}
 	return ws, nil
 }

--- a/integration/cluster.go
+++ b/integration/cluster.go
@@ -795,15 +795,7 @@ func (c *ClusterV3) Terminate(t *testing.T) {
 }
 
 func (c *ClusterV3) RandClient() *clientv3.Client {
-	for i := 0; i < 100; i++ {
-		cli := c.clients[rand.Intn(len(c.clients))]
-		if cli.ActiveConnection() == nil {
-			time.Sleep(10 * time.Millisecond)
-			continue
-		}
-		return cli
-	}
-	panic("failed to get a active client")
+	return c.clients[rand.Intn(len(c.clients))]
 }
 
 func (c *ClusterV3) Client(i int) *clientv3.Client {

